app/models/authenticator/httpd.rb

```diff
@@ -55,15 +55,52 @@ def update_user_attributes(user, _username, identity)
       user.name       = user.userid if user.name.blank?
     end
 
+    def find_or_initialize_user(identity, username)
+      user_attrs, _membership_list = identity
+      return super if user_attrs[:domain].nil?
+
+      upn_username = "#{user_attrs[:username]}@#{user_attrs[:domain]}".downcase
+
+      user = find_userid_as_upn(upn_username)
+      user ||= find_userid_as_distinguished_name(user_attrs, upn_username)
+      user ||= find_userid_as_username(identity, username, upn_username)
+      user ||= User.new(:userid => upn_username)
+
+      [upn_username, user]
+    end
+
     private
 
+    def find_userid_as_upn(upn_username)
+      user = User.find_by_userid(upn_username)
+      user || User.in_my_region.where('lower(userid) = ?', upn_username).order(:lastlogon).last
+    end
+
+    def find_userid_as_username(identity, username, upn_username)
+      userid = userid_for(identity, username)
+      user   = User.find_by_userid(userid)
+      user ||= User.in_my_region.where('lower(userid) = ?', userid).order(:lastlogon).last
+      $audit_log.info("Updating userid from #{user.userid} to #{upn_username}") unless user.blank?
+
+      user
+    end
+
+    def find_userid_as_distinguished_name(user_attrs, upn_username)
+      dn_domain = user_attrs[:domain].downcase.split(".").map { |s| "dc=#{s}" }.join(",")
+      user = User.in_my_region.where("userid LIKE ?", "%=#{user_attrs[:username]},%,#{dn_domain}").last
+      $audit_log.info("Updating userid from #{user.userid} to #{upn_username}") unless user.blank?
```

Comment:
Is this really "Updating userid..."? It's not clear where it's setting it and saving it.